source/slang/slang-ir-variable-scope-correction.cpp

```diff
@@ -0,0 +1,139 @@
+
+    Dictionary<IRInst*, List<IRInst*>> workListMap;
+
+    // traverse each instruction in the function
+    for (auto block = funcInst->getFirstBlock(); block; block = block->getNextBlock())
+    {
+        for (auto inst = block->getFirstChild(); inst; inst = inst->getNextInst())
+        {
+            uint32_t indent = 1;
+            printf("process inst = %d\n", inst->getOp());
+            // traverse the dominator tree, theoretically, the variables in each of dominator should be accessible in the current block,
+            // except the loop block, so we have to find out if there is a loop block in the dominator chain.
+            auto dominatorBlock = dominatorTree->getImmediateDominator(block);
+            for (; dominatorBlock; dominatorBlock = dominatorTree->getImmediateDominator(dominatorBlock))
+            {
+                debugPrint(indent++, dominatorBlock, "dominator block");
+                if (auto loop = as<IRLoop>(dominatorBlock->getTerminator()))
+                {
+                    debugPrint(indent, loop, "loop block");
+                    // Get the break block of the loop and check if such block
+                    auto breakBlock = loop->getBreakBlock();
+                    debugPrint(indent++, breakBlock, "break block");
+
+                    // If a block is not dominated by the break block, but dominated by the loop header, then the
+                    // instructions in this block are considered to be defined in the loop. We need to search
+                    // the instruction's use sites to see if there is any uses are out of the loop.
+                    if (!dominatorTree->dominates(breakBlock, block))
+                    {
+                        debugPrint(indent++, block, "block is not dominated by break block");
+                        // traverse all uses of this instruction
+                        for (auto use = inst->firstUse; use; use=use->nextUse)
+                        {
+                            debugPrint(indent++, use->getUser()->getParent(), "inst's use block is");
+                            if (auto userBlock = as<IRBlock>(use->getUser()->getParent()))
+                            {
+                                // If the use site of this instruction is dominated by the break block, it means that the
+                                // instruction is used after the break block, so we need to make that instruction available globally.
+                                // By doing so, we record all the users of this instructions.
+                                if (dominatorTree->dominates(breakBlock, userBlock))
+                                {
+                                    debugPrint(indent, inst, "inst is defined in loop but used outside of loop");
+                                    debugPrint(indent++, use->getUser(), "The user is");
+
+                                    auto list = workListMap.tryGetValue(inst);
```

Comment:
workListMap.add(inst) returns true if insts does not exist and was added by the call. Use this to avoid querying the hash map twice.

Reply:
Done.